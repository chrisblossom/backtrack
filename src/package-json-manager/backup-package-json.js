/* @flow */

import path from 'path';
import { get, isEqual } from 'lodash';
import { backupFile } from '../utils/backup-file';
import { rootPath } from '../config/paths';
import log from '../utils/log';
import { mapObjectKeyNames } from '../utils/object-utils';

import type { PackageJson } from '../types.js';

type Return = Promise<?{
    file: string,
    filename: string,
}>;

const npmDefaults = {
    version: '1.0.0',
    description: '',
    main: 'index.js',
    scripts: {
        test: 'echo "Error: no test specified" && exit 1',
    },
    keywords: [],
    author: '',
    license: 'ISC',
};

async function backupPackageJson(
    packageJson: PackageJson,
    managedKeys: ?PackageJson,
    previousManagedKeys: ?PackageJson,
): Return {
    const mapManagedKeys = mapObjectKeyNames(managedKeys);
    /**
     * backup package.json if any custom managed scripts exist
     */
    const shouldBackup = mapManagedKeys.some((key) => {
        const matchedMangedKey = get(managedKeys, key);
        const matchedPackageJson = get(packageJson, key);
        const matchedPreviousManagedKey = get(previousManagedKeys, key);
        const matchedDefaultKey = get(npmDefaults, key);

        /**
         * Don't backup if package.json key === npm's default
         */
        if (
            matchedPreviousManagedKey === undefined &&
            isEqual(matchedPackageJson, matchedDefaultKey)
        ) {
            return false;
        }

        // use isEqual to match arrays
        const result =
            matchedPackageJson &&
            isEqual(matchedMangedKey, matchedPackageJson) === false &&
            isEqual(matchedPreviousManagedKey, matchedPackageJson) === false;

        return result;
    });

    if (shouldBackup === false) {
        return undefined;
    }

    const packageJsonFile = path.resolve(rootPath, 'package.json');
    const backupData = await backupFile(packageJsonFile);

    if (!backupData) {
        return undefined;
    }

    const relativeBackupFilePath = path.relative(rootPath, backupData.file);

    log.warn(
        `Unknown package.json scripts. Backing up to: ${relativeBackupFilePath}`,
    );

    return {
        file: backupData.file,
        filename: backupData.filename,
    };
}

export { backupPackageJson };
