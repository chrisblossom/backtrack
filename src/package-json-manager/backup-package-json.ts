import path from 'path';
import { get, isEqual } from 'lodash';
import { backupFile } from '../utils/backup-file';
import { rootPath } from '../config/paths';
import log from '../utils/log';
import { mapObjectKeyNames } from '../utils/object-utils';

import { PackageJson } from '../types';

type Return = Promise<{
    file: string;
    filename: string;
} | void>;

const npmDefaults = {
    version: '1.0.0',
    description: '',
    main: 'index.js',
    scripts: {
        test: 'echo "Error: no test specified" && exit 1',
    },
    keywords: [],
    author: '',
    license: 'ISC',
};

async function backupPackageJson(
    packageJson: PackageJson,
    managedKeys?: PackageJson,
    previousManagedKeys?: PackageJson,
): Return {
    const mapManagedKeys = mapObjectKeyNames(managedKeys);
    /**
     * backup package.json if any custom managed scripts exist
     */

    const shouldBackup = mapManagedKeys
        .filter((key) => {
            const matchedMangedKey = get(managedKeys, key);
            const matchedPackageJson = get(packageJson, key);
            const matchedPreviousManagedKey = get(previousManagedKeys, key);
            const matchedDefaultKey = get(npmDefaults, key);

            /**
             * Don't backup if package.json key === npm default
             */
            if (
                matchedPreviousManagedKey === undefined &&
                isEqual(matchedPackageJson, matchedDefaultKey)
            ) {
                return false;
            }

            // use isEqual to match arrays
            const result =
                matchedPackageJson &&
                isEqual(matchedMangedKey, matchedPackageJson) === false &&
                isEqual(matchedPreviousManagedKey, matchedPackageJson) ===
                    false;

            return result;
        })
        .map((backupKeys) => {
            return backupKeys.join('.');
        });

    if (shouldBackup.length === 0) {
        return undefined;
    }

    const packageJsonFile = path.resolve(rootPath, 'package.json');
    const backupData = await backupFile(packageJsonFile);

    if (!backupData) {
        return undefined;
    }

    const relativeBackupFilePath = path.relative(rootPath, backupData.file);

    log.warn(
        `Unknown package.json keys: "${shouldBackup.join(
            ', ',
        )}". Backing up to: ${relativeBackupFilePath}`,
    );

    return {
        file: backupData.file,
        filename: backupData.filename,
    };
}

export { backupPackageJson };
