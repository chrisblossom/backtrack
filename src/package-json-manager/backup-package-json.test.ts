import path from 'path';

const backupPackageJson = (
	packageJson: any,
	managedKeys: any,
	previousManagedKeys?: any,
) =>
	require('./backup-package-json').backupPackageJson(
		packageJson,
		managedKeys,
		previousManagedKeys,
	);

describe('backupPackageJson', () => {
	const cwd = process.cwd();
	let move: any;

	beforeEach(() => {
		jest.spyOn(console, 'info').mockImplementation(() => undefined);
		jest.spyOn(console, 'warn').mockImplementation(() => undefined);
		jest.spyOn(console, 'error').mockImplementation(() => undefined);
		jest.spyOn(console, 'debug').mockImplementation(() => undefined);

		move = jest.requireMock('fs-extra').move;

		jest.mock('fs-extra', () => ({
			move: jest.fn(() => Promise.resolve()),
		}));

		/**
		 * Change dir to keep consistent package.json hash for filename
		 */
		const dir = path.resolve(__dirname, '__sandbox__/package-json-1/');
		process.chdir(dir);
	});

	afterEach(() => {
		process.chdir(cwd);
	});

	test('backup package.json when a managed script already exists', async () => {
		const packageJson = {
			name: 'testing',
			scripts: {
				start: 'start',
				test: 'jest',
			},
		};

		const managedKeys = {
			scripts: {
				test: 'backtrack test',
				lint: 'backtrack lint',
			},
		};

		const previousManagedKeys = {
			scripts: {
				test: 'backtrack test',
			},
		};

		const backupResult = await backupPackageJson(
			packageJson,
			managedKeys,
			previousManagedKeys,
		);

		expect(move.mock.calls).toMatchSnapshot();
		expect(backupResult).toMatchSnapshot();
	});

	test('backup package.json when multiple keys already exist', async () => {
		const packageJson = {
			name: 'testing',
			main: 'fake-entry.js',
			scripts: {
				start: 'start',
				test: 'jest',
			},
		};

		const managedKeys = {
			main: 'real-entry.js',
			scripts: {
				test: 'backtrack test',
				lint: 'backtrack lint',
			},
		};

		const previousManagedKeys = {
			scripts: {
				test: 'backtrack test',
			},
		};

		const backupResult = await backupPackageJson(
			packageJson,
			managedKeys,
			previousManagedKeys,
		);

		expect(move.mock.calls).toMatchSnapshot();
		expect(backupResult).toMatchSnapshot();
	});

	test('does not backup if script does not exist yet', async () => {
		const packageJson = {
			name: 'testing',
			scripts: {
				start: 'start',
			},
		};

		const managedKeys = {
			scripts: {
				test: 'backtrack test',
				lint: 'backtrack lint',
			},
		};

		const previousManagedKeys = {
			scripts: {
				test: 'backtrack test',
			},
		};

		const backupResult = await backupPackageJson(
			packageJson,
			managedKeys,
			previousManagedKeys,
		);

		expect(move.mock.calls).toEqual([]);
		expect(backupResult).toEqual(undefined);
	});

	test('does not backup if script equals previousManaged', async () => {
		const packageJson = {
			name: 'testing',
			scripts: {
				start: 'start',
				lint: 'previous',
			},
		};

		const managedKeys = {
			scripts: {
				test: 'backtrack test',
				lint: 'backtrack lint',
			},
		};

		const previousManagedKeys = {
			scripts: {
				test: 'backtrack test',
				lint: 'previous',
			},
		};

		const backupResult = await backupPackageJson(
			packageJson,
			managedKeys,
			previousManagedKeys,
		);

		expect(move.mock.calls).toEqual([]);
		expect(backupResult).toEqual(undefined);
	});

	test('works with array value - no backup', async () => {
		const packageJson = {
			name: 'testing',
			'lint-staged': {
				'*.{js,md,json}': [
					'prettier --write',
					'git add',
				],
			},
		};

		const managedKeys = {
			'lint-staged': {
				'*.{js,md,json}': [
					'prettier --write',
					'git add',
				],
			},
		};

		const previousManagedKeys = {
			'lint-staged': {
				'*.{js,md,json}': [
					'prettier --write',
					'git add',
				],
			},
		};

		const backupResult = await backupPackageJson(
			packageJson,
			managedKeys,
			previousManagedKeys,
		);

		expect(move.mock.calls).toEqual([]);
		expect(backupResult).toEqual(undefined);
	});

	test('works with array value - backup', async () => {
		const packageJson = {
			name: 'testing',
			'lint-staged': {
				'*.{js,md,json}': ['git add'],
			},
		};

		const managedKeys = {
			'lint-staged': {
				'*.{js,md,json}': [
					'prettier --write',
					'git add',
				],
			},
		};

		const previousManagedKeys = {
			'lint-staged': {
				'*.{js,md,json}': [
					'prettier --write',
					'git add',
				],
			},
		};

		const backupResult = await backupPackageJson(
			packageJson,
			managedKeys,
			previousManagedKeys,
		);

		expect(move.mock.calls).toMatchSnapshot();
		expect(backupResult).toMatchSnapshot();
	});

	test('handles npm default scripts', async () => {
		const packageJson = {
			name: 'testing',
			keywords: [],
			scripts: {
				start: 'start',
				test: 'echo "Error: no test specified" && exit 1',
			},
		};

		const managedKeys = {
			scripts: {
				test: 'backtrack test',
				lint: 'backtrack lint',
			},
			keywords: ['keyword'],
		};

		const backupResult = await backupPackageJson(packageJson, managedKeys);

		expect(move.mock.calls).toEqual([]);
		expect(backupResult).toEqual(undefined);
	});

	test('handles npm default test script with previously managed', async () => {
		const packageJson = {
			name: 'testing',
			scripts: {
				start: 'start',
				test: 'echo "Error: no test specified" && exit 1',
			},
		};

		const managedKeys = {
			scripts: {
				test: 'backtrack test',
				lint: 'backtrack lint',
			},
		};

		const previousManagedKeys = {
			scripts: {
				test: 'backtrack test',
				lint: 'backtrack lint',
			},
		};

		const backupResult = await backupPackageJson(
			packageJson,
			managedKeys,
			previousManagedKeys,
		);

		expect(move.mock.calls).toMatchSnapshot();
		expect(backupResult).toMatchSnapshot();
	});
});
