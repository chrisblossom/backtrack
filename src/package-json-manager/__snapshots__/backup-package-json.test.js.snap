// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`backupPackageJson backup package.json when a managed script already exists 1`] = `
Array [
  Array [
    "<PROJECT_ROOT>/package.json",
    "<PROJECT_ROOT>/package.json-backup-a2b834f3.json",
    Object {
      "overwrite": false,
    },
  ],
]
`;

exports[`backupPackageJson backup package.json when a managed script already exists 2`] = `
Object {
  "file": "<PROJECT_ROOT>/package.json-backup-a2b834f3.json",
  "filename": "package.json-backup-a2b834f3.json",
}
`;

exports[`backupPackageJson handles npm default test script with previously managed 1`] = `
Array [
  Array [
    "<PROJECT_ROOT>/package.json",
    "<PROJECT_ROOT>/package.json-backup-a2b834f3.json",
    Object {
      "overwrite": false,
    },
  ],
]
`;

exports[`backupPackageJson handles npm default test script with previously managed 2`] = `
Object {
  "file": "<PROJECT_ROOT>/package.json-backup-a2b834f3.json",
  "filename": "package.json-backup-a2b834f3.json",
}
`;

exports[`backupPackageJson works with array value - backup 1`] = `
Array [
  Array [
    "<PROJECT_ROOT>/package.json",
    "<PROJECT_ROOT>/package.json-backup-a2b834f3.json",
    Object {
      "overwrite": false,
    },
  ],
]
`;

exports[`backupPackageJson works with array value - backup 2`] = `
Object {
  "file": "<PROJECT_ROOT>/package.json-backup-a2b834f3.json",
  "filename": "package.json-backup-a2b834f3.json",
}
`;
